### PR TITLE
Add bsr_dense_addmm triton kernel

### PR DESCRIPTION
**Pull request by pearu**
As in the title.

The `bsr_dense_addmm` kernel implemented in this PR is a generalization of `bsr_dense_mm` in the following respects (in addition of having input, beta, and alpha parameters):
- it implements `SPLIT_N` kernel parameter that enables efficient kernel launches in the case of wide inputs. For instance, the timing of nn.linear with 256x256 BSR weights having 16x16 blocks and 256x131072 strided input reduced about 16x (this corresponds to the 94 % speed up value listed below).
- it supports rectangular blocks in sparse BSR tensor weights

The performance increase of nn.linear is as follows (float16, `NVIDIA A100-SXM4-80GB`):
- with 16x16 blocks, the average/maximal speed up is  55/94 %
- with 32x32 blocks, the average/maximal speed up is  33/63 %
- with 64x64 blocks, the average/maximal speed up is  23/42 %
- with 128x128 blocks, the average/maximal speed up is  15/39 %

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114595

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -1274,6 +1274,12 @@ def kernel_impl(*args, **kwargs):
 
         return bsr_dense_mm(*args, skip_checks=True, **kwargs)
 
+    @_WrappedTritonKernel
+    def addmm_kernel_impl(*args, **kwargs):
+        from torch.sparse._triton_ops import bsr_dense_addmm
```
Comment (vadimkantorov) [inline torch/cuda/__init__.py:1279]: I've noticed once that doing import in the hot-path is still somewhat slow, even if the module is imported already. Hope this is not the case here.

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**